### PR TITLE
Dossiertemplates add wizard

### PR DESCRIPTION
**Pull request by elioschmutz**
Dieser PR fügt den Add-Wizard für "Dossier aus Template" hinzu:

screen3

ACHTUNG: Dieser PR behandelt nicht das rekursive erstellen von Dossiers und Dokumenten. Es geht hier lediglich um den Wizard und um das initiale Erstellen eines Basisdossiers aufgrund der Template-Attribute.

see #2143

### CODE REVIEW

**Comment by elioschmutz** [comment]
@phgross ready to review

File: opengever/dossier/templatedossier/__init__.py [hunk 1]
```diff
@@ -12,3 +15,28 @@ def get_template_dossier():
     if result:
         return result[0].getObject()
     return None
+
+
+def get_templates(template_type, depth=-1):
+    template_dossier = get_template_dossier()
+
+    if template_dossier is None:
```
Comment (phgross) [inline opengever/dossier/templatedossier/__init__.py]: Fetching the `template dossier` is unnecessary, just make a catalog query and get all dosisertemplates, so the generalization of the `get_templates` can be reverted.

File: opengever/dossier/tests/test_dossier_template.py [hunk 2]
```diff
@@ -233,3 +236,116 @@ def test_is_available_on_leaf_node_when_feature_is_enabled(self):
 
         self.assertTrue(
             leaf_node.restrictedTraverse('@@dossier_with_template/is_available')())
+
+    @browsing
+    def test_only_show_dossiertemplates_without_subdossiers(self, browser):
+        templatedossier = create(Builder('templatedossier'))
+
+        root = create(Builder('repository_root'))
```
Comment (phgross) [inline opengever/dossier/tests/test_dossier_template.py]: The repository root and maybe other content too are used in every test, move them to the setUp method.

File: opengever/dossier/tests/test_dossier_template.py [hunk 3]
```diff
@@ -233,3 +236,116 @@ def test_is_available_on_leaf_node_when_feature_is_enabled(self):
...
+        browser.login().visit(leaf_node)
+        factoriesmenu.add('Dossier with template')
+
+        token = browser.css(
+            'input[name="form.widgets.template"]').first.attrib.get('value')
+
+        browser.fill({'form.widgets.template': token}).submit()
+
+        browser.fill({'Responsible': TEST_USER_ID})
+        browser.find('Save').click()
```
Comment (phgross) [inline opengever/dossier/tests/test_dossier_template.py]: There is a shortcut `browser.click_on('Save')`

File: opengever/dossier/tests/test_dossier_template.py [hunk 4]
```diff
@@ -233,3 +236,116 @@ def test_is_available_on_leaf_node_when_feature_is_enabled(self):
...
+
+        dossier = browser.context
+        self.assertEqual(
```
Comment (phgross) [inline opengever/dossier/tests/test_dossier_template.py]: I guess this test will never fail. 😉 

I would made the following assertion:
```python
self.assertEqual(leaf_node('dossier-1'), browser.context)
```

Besides that I would extend this test with an assertion on the statusmessage.

File: opengever/dossier/tests/test_dossier_template.py [hunk 5]
```diff
@@ -233,3 +236,116 @@ def test_is_available_on_leaf_node_when_feature_is_enabled(self):
...
+        dossier = browser.context
+
+        self.assertEqual(values.get('title'), dossier.title)
+        self.assertEqual(values.get('description'), dossier.description)
+        self.assertEqual(values.get('keywords'), IDossier(dossier).keywords)
+        self.assertEqual(values.get('comments'), IDossier(dossier).comments)
+        self.assertEqual(values.get('filing_prefix'), IDossier(dossier).filing_prefix)
+
+    @browsing
+    def test_redirect_to_dossier_after_creating_dossier_from_template(self, browser):
```
Comment (phgross) [inline opengever/dossier/tests/test_dossier_template.py]: redirect**s**

File: opengever/dossier/tests/test_dossier_template.py [hunk 6]
```diff
@@ -233,3 +236,116 @@ def test_is_available_on_leaf_node_when_feature_is_enabled(self):
...
+
+        autocomplete_selection_widget = leaf_node.unrestrictedTraverse(
+            'add-dossier-template/++widget++IDossier.responsible')
+
+        self.assertIsInstance(
+            autocomplete_selection_widget,
+            AutocompleteSelectionWidget)
+
+    @browsing
+    def test_redirect_to_context_if_the_user_skips_the_first_wizard_step(self, browser):
```
Comment (phgross) [inline opengever/dossier/tests/test_dossier_template.py]: redirect**s**

File: opengever/dossier/dossiertemplate/form.py [hunk 7]
```diff
@@ -1,10 +1,209 @@
...
+    steps = [
+        ('select-dossier', _(u'Select dossier')),
+        ('add-dossier-template', _(u'Add dossier'))]
+
+
+class SelectDossierTemplateWizardStep(
+        CreateDossierMixin, AutoExtensibleForm, BaseWizardStepForm, Form):
+    """First wizard step - select dossiertemplate
+    """
+    step_name = 'select-dossier'
```
Comment (phgross) [inline opengever/dossier/dossiertemplate/form.py]: `select-template`

File: opengever/dossier/dossiertemplate/form.py [hunk 8]
```diff
@@ -1,10 +1,209 @@
...
+
+    def updateFieldsFromSchemata(self):
```
Comment (phgross) [inline opengever/dossier/dossiertemplate/form.py]: copy & paste --> ✂️

File: opengever/dossier/dossiertemplate/form.py [hunk 9]
```diff
@@ -1,10 +1,209 @@
...
+        set_wizard_storage(self.context, data)
+        return self.request.RESPONSE.redirect(
+            "{}/add-dossier-template".format(self.context.absolute_url()))
+
+    @button.buttonAndHandler(_(u'button_cancel', default=u'Cancel'), name='cancel')
+    def cancel(self, action):
+        return self.request.RESPONSE.redirect(self.context.absolute_url())
+
+
+class AddDossierTemplateWizardStep(WizzardWrappedAddForm):
```
Comment (phgross) [inline opengever/dossier/dossiertemplate/form.py]: The name is wrong or confusing, the user don't add a dossier template.

File: opengever/dossier/dossiertemplate/form.py [hunk 10]
```diff
@@ -1,10 +1,209 @@
...
+    """Second wizard step - add dossiertemplate
+    """
+    grok.context(IRepositoryFolder)
+    grok.name('add-dossier-template')
```
Comment (phgross) [inline opengever/dossier/dossiertemplate/form.py]: same here

File: opengever/dossier/dossiertemplate/form.py [hunk 11]
```diff
@@ -1,10 +1,209 @@
...
+
+    typename = 'opengever.dossier.businesscasedossier'
+
+    def _create_form_class(self, parent_form_class, steptitle):
+        class WrappedForm(CreateDossierMixin, BaseWizardStepForm, parent_form_class):
+            step_name = 'add-dossier-template'
+
+            def update(self):
+                """Here we update the widget-values of the dossier add-form
```
Comment (phgross) [inline opengever/dossier/dossiertemplate/form.py]: ~~Here we~~ Update wigets-values with the values of the selected dossiertemplate.

File: opengever/dossier/dossiertemplate/form.py [hunk 12]
```diff
@@ -1,10 +1,209 @@
...
+                if not getSecurityManager().getUser().getId():
+                    # this happens during ++widget++ traversal
+                    return
+
+                template_obj = get_wizard_storage(self.context).get('template')
+
+                if not template_obj:
+                    # This happens if the user access the step directly and
+                    # the wizard storage was expired or never existent.
+                    return self.request.RESPONSE.redirect(self.context.absolute_url())
```
Comment (phgross) [inline opengever/dossier/dossiertemplate/form.py]: shows a statusmessage

File: opengever/dossier/dossiertemplate/form.py [hunk 13]
```diff
@@ -1,10 +1,209 @@
...
+                        # Skip not whitelisted template fields.
+                        # We don't want to update fields which are not
+                        # whitelisted in the template.
+                        template_widget_name = self.map_widget_names_interfaces(widgetname)
+                        if template_widget_name not in TEMPLATABLE_FIELDS:
+                            continue
+
+                        value = template_values.get(template_widget_name)
+                        widget = group.widgets.get(widgetname)
+                        converter = IDataConverter(widget)
```
Comment (phgross) [inline opengever/dossier/dossiertemplate/form.py]: simplify:

```python
widget.value = IDataConverter(widget).toWidgetValue(value)
```

File: opengever/dossier/dossiertemplate/form.py [hunk 14]
```diff
@@ -1,10 +1,209 @@
...
+
+                        # Set the template value to the dossier add-form widget.
+                        widget.value = converter.toWidgetValue(value)
+
+            def map_widget_names_interfaces(self, widgetname):
```
Comment (phgross) [inline opengever/dossier/dossiertemplate/form.py]: get_template_widget_name(self, widgetname)

File: opengever/dossier/dossiertemplate/form.py [hunk 15]
```diff
@@ -1,10 +1,209 @@
...
+                """The dossiertemplate uses the same fields like for i.e. the
```
Comment (phgross) [inline opengever/dossier/dossiertemplate/form.py]: `The dossiertemplate uses the same fields as the dossier (IDossier) but it includes it with another interface. We have to map this two interfaces names to get the correct value or widget.`

**Comment by elioschmutz** [comment]
@phgross updated

File: opengever/dossier/tests/test_dossier_template.py [hunk 16]
```diff
@@ -210,26 +213,115 @@ class TestDossierTemplateAddWizard(FunctionalTestCase):
...
+
+        browser.login().visit(self.leaf_node)
+        factoriesmenu.add('Dossier with template')
+
+        token = browser.css(
+            'input[name="form.widgets.template"]').first.attrib.get('value')
+
+        browser.fill({'form.widgets.template': token}).submit()
+
+        browser.fill({'Responsible': TEST_USER_ID})
```
Comment (phgross) [inline opengever/dossier/tests/test_dossier_template.py]: The test for the default value ist still missing. Filling in the `responsible` should not be necessary.

**Comment by elioschmutz** [comment]
@phgross 👍 , and the test for default values is here: https://github.com/4teamwork/opengever.core/pull/2356/files#diff-79c1083ed632c03164d578fbbaa4f0e1R426

**Pull request merged**